### PR TITLE
LDAP to not register new users when outside of fair use or over limits

### PR DESCRIPTION
**Pull request by blizzz**
In `UserManager::createUser` we check whether the instance is over limits (subscription or one-click-instance) and would refuse to add users. It works for local backends only. This PR offers a comfort method for assertion and makes the LDAP backend taking advantage of it.

### CODE REVIEW

**Comment by blizzz** [comment]
Downside of the current solution:  With on-demand mapping, an end user may face an infinite spinner when trying to share a file, when a user would have to be mapped. The sharing frontend code does not handle well 500 results. 

My idea now is to perhaps react on the request: throw the exception only when the request is done via provisioning api, otherwise just skip mapping.

File: lib/private/Support/Subscription/Assertion.php [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+		$this->notificationManager = $notificationManager;
+	}
+
+	/**
+	 * @inheritDoc
+	 */
+	public function createUserIsLegit(): void {
+		if ($this->registry->delegateIsHardUserLimitReached($this->notificationManager)) {
+			$l = $this->l10nFactory->get('lib');
+			throw new HintException($l->t('The user limit has been reached and the user was not created.'));
```
Comment (szaimen) [inline lib/private/Support/Subscription/Assertion.php:52]: to comply with https://github.com/nextcloud/server/pull/34706

```suggestion
			throw new HintException($l->t('Nextcloud Enterprise subscription needed for more users. Check your notifications to learn more.'));
```

Comment (blizzz) [reply lib/private/Support/Subscription/Assertion.php:52]: It might be shown as well when a subscription exists, but is exceeded iirc.

Comment (szaimen) [reply lib/private/Support/Subscription/Assertion.php:52]: all right, then lets leave it like this 👍

File: apps/user_ldap/lib/Jobs/CleanUp.php [hunk 2]
```diff
@@ -107,7 +107,7 @@ public function setArguments($arguments): void {
 		if (isset($arguments['mapping'])) {
 			$this->mapping = $arguments['mapping'];
 		} else {
-			$this->mapping = new UserMapping($this->db);
+			$this->mapping = \OCP\Server::get(UserMapping::class);
```
Comment (PVince81) [inline apps/user_ldap/lib/Jobs/CleanUp.php:110]: no DI possible I guess or too expensive ?

File: apps/user_ldap/lib/Mapping/UserMapping.php [hunk 3]
```diff
@@ -22,12 +22,51 @@
...
+
 /**
  * Class UserMapping
+ *
  * @package OCA\User_LDAP\Mapping
  */
 class UserMapping extends AbstractMapping {
 
+	private IAssertion $assertion;
+	protected const PROV_API_REGEX = '/\/ocs\/v[1-9].php\/cloud\/(groups|users)/';
```
Comment (PVince81) [inline apps/user_ldap/lib/Mapping/UserMapping.php:39]: would this also be possible via a Middleware? not sure which approach is best

File: apps/user_ldap/lib/Jobs/CleanUp.php (hunk 2 above)
Comment (blizzz) [reply apps/user_ldap/lib/Jobs/CleanUp.php:110]: Despite being a low hanging fruit here, I do not intend to max out refactoring, with backporting in mind.

File: apps/user_ldap/lib/Mapping/UserMapping.php (hunk 3 above)
Comment (blizzz) [reply apps/user_ldap/lib/Mapping/UserMapping.php:39]: maybe, this crossed my mind, but did not follow that path :watch:

**Review by PVince81 — APPROVED** [review]
:+1: code looks fine

**Review by szaimen — APPROVED** [review]
🐘

**Comment by szaimen** [comment]
Can someone restart drone? Not sure if the failure is really related...

**Comment by PVince81** [comment]
yeah, a bit weird:
```
    The Language column in Users Settings could not be found after 100 seconds (NoSuchElementException)
```

I've restarted the Drone builds

**Pull request merged**